File: app/src/main/java/com/getman/varnabeach/recycler/BeachAdapter.java
package com.getman.varnabeach.recycler;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;

import com.getman.varnabeach.room.Beach;

public class BeachAdapter extends ListAdapter<Beach, BeachViewHolder> {

    public BeachAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public BeachViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return BeachViewHolder.create(parent);
    }

    @Override
    public void onBindViewHolder(@NonNull BeachViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    private static final DiffUtil.ItemCallback<Beach> DIFF_CALLBACK = new DiffUtil.ItemCallback<Beach>() {
        @Override
        public boolean areItemsTheSame(@NonNull Beach oldItem, @NonNull Beach newItem) {
            return oldItem.id == newItem.id;
        }

        @Override
        public boolean areContentsTheSame(@NonNull Beach oldItem, @NonNull Beach newItem) {
            return oldItem.equals(newItem);
        }
    };
}

File: app/src/main/java/com/getman/varnabeach/recycler/BeachAdapter.kt
package com.getman.varnabeach.recycler

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.getman.varnabeach.recycler.BeachViewHolder.Companion.create
import com.getman.varnabeach.room.Beach

class BeachAdapter(
    private val onItemClicked: (Beach) -> Unit
) : ListAdapter<Beach, BeachViewHolder>(DIFF_CALLBACK()) {

    private class DIFF_CALLBACK : DiffUtil.ItemCallback<Beach>() {
        override fun areItemsTheSame(oldItem: Beach, newItem: Beach): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Beach, newItem: Beach): Boolean {
            return oldItem == newItem
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BeachViewHolder {
        return create(parent) {
            onItemClicked(getItem(it))
        }
    }

    override fun onBindViewHolder(holder: BeachViewHolder, position: Int) {
        holder.bind(getItem(position)!!)
    }
}